pkg/vault/contracts/Vault.sol

```diff
@@ -1234,40 +1247,29 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
-                    // amount in needs to be discounted for that. For EXACT_OUT, `vaultWrappedDeltaHint` and
-                    // `amountOutWrapped` do not change after the mint operation, so the `bufferWrappedSurplus` does
-                    // not need to be recalculated.
-                    bufferUnderlyingSurplus = wrappedToken.previewMint(bufferWrappedSurplus);
-                    amountInUnderlying = vaultUnderlyingDeltaHint - bufferUnderlyingSurplus;
-                }
-
-                // In a wrap operation, the underlying balance of the buffer will decrease and the wrapped balance will
-                // increase. To decrease the underlying balance, we get the delta amount that was deposited
-                // (vaultUnderlyingDeltaHint) and discount the amount needed for the wrapping operation
-                // (amountInUnderlying). The same logic applies to wrapped balances.
-                //
-                // Note: bufferUnderlyingSurplus = vaultUnderlyingDeltaHint - amountInUnderlying
-                //       bufferWrappedSurplus = vaultWrappedDeltaHint - amountOutWrapped
-                bufferBalances = PackedTokenBalance.toPackedBalance(
-                    bufferBalances.getBalanceRaw() - bufferUnderlyingSurplus,
-                    bufferBalances.getBalanceDerived() + bufferWrappedSurplus
-                );
-                _bufferTokenBalances[wrappedToken] = bufferBalances;
-            } else {
-                amountInUnderlying = vaultUnderlyingDeltaHint;
-                amountOutWrapped = vaultWrappedDeltaHint;
-            }
+            // A wrap should remove underlying from the sender and buffer, and return wrapped tokens to the sender
+            // and buffer. However, `vaultUnderlyingDeltaHint` may be smaller than `amountInUnderlying`, meaning
+            // that the buffer had more wrapped than underlying. If that's the case, an amount of underlying must
+            // be added to the buffer, so `bufferUnderlyingImbalance` must be negative (note that the buffer
+            // balance update below already removes bufferUnderlyingImbalance from the buffer, so if this number
+            // is negative it'll actually be added to the buffer balances).
+            int256 bufferUnderlyingImbalance = vaultUnderlyingDeltaHint.toInt256() - amountInUnderlying.toInt256();
+            // A wrap should remove underlying from the sender and buffer, and return wrapped tokens to the sender
+            // and buffer. However, `vaultWrappedDeltaHint` may be smaller than `amountOutWrapped`, meaning
+            // that the buffer had more wrapped than underlying. If that's the case, an amount of wrapped must
+            // be removed from the buffer, so `bufferWrappedImbalance` must be negative.
+
+            int256 bufferWrappedImbalance = vaultWrappedDeltaHint.toInt256() - amountOutWrapped.toInt256();
```

Comment:
last minute megasilly comment, but it's something that helps me follow the code better: the comments in the code are good. But more than "imbalance" this is some sort of `netBufferUnderlyingDelta` / `netBufferWrappedDelta`, measured according to the real balances entering and exiting the vault. Or maybe `bufferOutgoingUnderlyingAmount`, `bufferIncomingWrappedAmount` if you want to preserve the signs when we adjust the balance.

Bottom line, this is not the buffer imbalance that we are calculating; we already do that before wrapping. We're calculating how much to take out from buffer underlying balances (i.e. whatever the vault used to wrap `vaultUnderlyingDeltaHint` minus whatever the user paid `amountInUnderlying`) and how much to add to buffer wrapped balances (i.e. whatever the vault received from the lending protocol `vaultWrappedDeltaHint` minus whatever the vault pays the user `amountOutWrapped`).

The buffer ends up rebalanced because of the math we did before wrapping; now we're just accounting for those results.

Reply:
Ok, I'll change the naming